File: frontend/types/astrology.ts
// Canonical Vedic astrology type definitions.
// All domain types live here. Other modules import from this file.

// ─── Enums (mirrored from backend domain) ────────────────────────────────────

export enum Ayanamsa {
  LAHIRI = "Lahiri",
  RAMAN = "Raman",
  KRISHNAMURTI = "Krishnamurti",
  TRUE_CHITRA = "True Chitra",
}

export enum DashaLord {
  KETU = "Ketu",
  VENUS = "Venus",
  SUN = "Sun",
  MOON = "Moon",
  MARS = "Mars",
  RAHU = "Rahu",
  JUPITER = "Jupiter",
  SATURN = "Saturn",
  MERCURY = "Mercury",
}

export enum DashaLevel {
  MAHADASHA = "Mahadasha",
  ANTARDASHA = "Antardasha",
  PRATYANTAR_DASHA = "Pratyantar Dasha",
  SUKSHMA_DASHA = "Sukshma Dasha",
  PRANA_DASHA = "Prana Dasha",
}

export enum ZodiacSign {
  ARIES = "Aries",
  TAURUS = "Taurus",
  GEMINI = "Gemini",
  CANCER = "Cancer",
  LEO = "Leo",
  VIRGO = "Virgo",
  LIBRA = "Libra",
  SCORPIO = "Scorpio",
  SAGITTARIUS = "Sagittarius",
  CAPRICORN = "Capricorn",
  AQUARIUS = "Aquarius",
  PISCES = "Pisces",
}

export enum Nakshatra {
  ASHWINI = "Ashwini",
  BHARANI = "Bharani",
  KRITTIKA = "Krittika",
  ROHINI = "Rohini",
  MRIGASHIRA = "Mrigashira",
  ARDRA = "Ardra",
  PUNARVASU = "Punarvasu",
  PUSHYA = "Pushya",
  ASHLESHA = "Ashlesha",
  MAGHA = "Magha",
  PURVA_PHALGUNI = "Purva Phalguni",
  UTTARA_PHALGUNI = "Uttara Phalguni",
  HASTA = "Hasta",
  CHITRA = "Chitra",
  SWATI = "Swati",
  VISHAKHA = "Vishakha",
  ANURADHA = "Anuradha",
  JYESHTHA = "Jyeshtha",
  MULA = "Mula",
  PURVA_ASHADHA = "Purva Ashadha",
  UTTARA_ASHADHA = "Uttara Ashadha",
  SHRAVANA = "Shravana",
  DHANISHTA = "Dhanishta",
  SHATABHISHA = "Shatabhisha",
  PURVA_BHADRAPADA = "Purva Bhadrapada",
  UTTARA_BHADRAPADA = "Uttara Bhadrapada",
  REVATI = "Revati",
}

export enum PlanetType {
  SUN = "Sun",
  MOON = "Moon",
  MARS = "Mars",
  MERCURY = "Mercury",
  JUPITER = "Jupiter",
  VENUS = "Venus",
  SATURN = "Saturn",
  RAHU = "Rahu",
  KETU = "Ketu",
}

// ─── Request types ────────────────────────────────────────────────────────────

export interface BirthChartRequest {
  date: string; // ISO 8601 format
  latitude: number;
  longitude: number;
  timezone: string;
  ayanamsa: Ayanamsa;
  house_system: number;
}

export interface AstrologyAnalysisRequest {
  date: string; // ISO 8601 format
  latitude: number;
  longitude: number;
  timezone: string;
  ayanamsa: Ayanamsa;
  house_system: number;
}

// ─── Response types ───────────────────────────────────────────────────────────

export interface PlanetPosition {
  planet: PlanetType;
  longitude: number;
  zodiac_sign: ZodiacSign;
  house_number: number;
  retrograde: boolean;
  nakshatra: Nakshatra;
  pada: number;
  degree_within_sign: number;
}

export interface HousePosition {
  house_number: number;
  start_longitude: number;
  end_longitude: number;
}

export interface AscendantPosition {
  zodiac_sign: ZodiacSign;
  longitude: number;
  nakshatra: Nakshatra;
  pada: number;
  degree_within_sign: number;
}

export interface BirthChartResponse {
  ascendant: AscendantPosition;
  planets: PlanetPosition[];
  houses: HousePosition[];
}

// Divisional charts share the same shape as BirthChartResponse.
export type NavamsaChartResponse = BirthChartResponse;
export type DasamsaChartResponse = BirthChartResponse;
export type ShastiamsaChartResponse = BirthChartResponse;

export interface PratyantarDashaResponse {
  lord: DashaLord;
  start_datetime: string;
  end_datetime: string;
  duration_days: number;
  level: DashaLevel;
}

export interface AntardashaResponse {
  lord: DashaLord;
  start_datetime: string;
  end_datetime: string;
  duration_days: number;
  level: DashaLevel;
  pratyantars: PratyantarDashaResponse[];
}

export interface MahadashaResponse {
  lord: DashaLord;
  start_datetime: string;
  end_datetime: string;
  duration_days: number;
  level: DashaLevel;
  antardashas: AntardashaResponse[];
}

export interface VimshottariDashaResponse {
  mahadashas: MahadashaResponse[];
}

export interface Yoga {
  name: string;
  description: string;
}

export interface AstrologyAnalysisResponse {
  birth_chart: BirthChartResponse;
  navamsa_chart: BirthChartResponse;
  dasamsa_chart: BirthChartResponse;
  shastiamsa_chart: BirthChartResponse;
  vimshottari_dashas: VimshottariDashaResponse;
  current_mahadasha?: string;
  current_antardasha?: string;
  detected_yogas: Yoga[];
  interpretation?: string;
  yoga_analysis?: string;
  career_score?: number;
  wealth_score?: number;
  relationship_score?: number;
  health_score?: number;
}

File: backend/app/schemas/astrology.py
from pydantic import BaseModel, Field, field_validator
from typing import List
from datetime import datetime
from app.domain.ayanamsa import Ayanamsa
from app.domain.dasha import DashaLevel, DashaLord
from app.domain.zodiac import ZodiacSign
from app.domain.nakshatra import Nakshatra
from app.domain.planet_type import PlanetType
from app.domain.house_number import HouseNumber


class BirthData(BaseModel):
    """Request schema for calculating a birth chart."""
    date: datetime = Field(..., description="Birth date and time in UTC")
    latitude: float = Field(..., ge=-90.0, le=90.0, description="Birth latitude")
    longitude: float = Field(..., ge=-180.0, le=180.0, description="Birth longitude")
    timezone: str = Field(..., description="IANA timezone identifier (e.g., 'Asia/Kolkata')")

    @field_validator("date")
    @classmethod
    def validate_timezone_aware_datetime(cls, value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("Birth date and time must include timezone information.")
        return value


class BirthChartRequest(BirthData):
    """Request schema for generating a birth chart."""
    ayanamsa: Ayanamsa = Field(..., description="Ayanamsa to use for sidereal calculation")
    house_system: int = Field(..., ge=1, description="House system identifier")


class PlanetPosition(BaseModel):
    """API response schema for a single planet's position."""
    planet: PlanetType
    longitude: float
    zodiac_sign: ZodiacSign
    house_number: HouseNumber
    retrograde: bool
    nakshatra: Nakshatra
    pada: int
    degree_within_sign: float


class HousePosition(BaseModel):
    """API response schema for a house's position."""
    house_number: HouseNumber
    start_longitude: float
    end_longitude: float


class AscendantPosition(BaseModel):
    """API response schema for the Ascendant."""
    zodiac_sign: ZodiacSign
    longitude: float
    nakshatra: Nakshatra
    pada: int
    degree_within_sign: float


class BirthChartResponse(BaseModel):
    """Complete API response for a birth chart."""
    ascendant: AscendantPosition
    planets: List[PlanetPosition]
    houses: List[HousePosition]


class NavamsaChartResponse(BirthChartResponse):
    """Complete API response for a Navamsa chart."""
    pass


class DasamsaChartResponse(BirthChartResponse):
    """Complete API response for a Dasamsa chart."""
    pass


class ShastiamsaChartResponse(BirthChartResponse):
    """Complete API response for a Shastiamsa chart."""
    pass


class PratyantarDashaResponse(BaseModel):
    """API response schema for a Pratyantar Dasha period."""
    lord: DashaLord
    start_datetime: datetime
    end_datetime: datetime
    duration_days: float
    level: DashaLevel


class AntardashaResponse(BaseModel):
    """API response schema for an Antardasha period."""
    lord: DashaLord
    start_datetime: datetime
    end_datetime: datetime
    duration_days: float
    level: DashaLevel
    pratyantars: List[PratyantarDashaResponse]


class MahadashaResponse(BaseModel):
    """API response schema for a Mahadasha period."""
    lord: DashaLord
    start_datetime: datetime
    end_datetime: datetime
    duration_days: float
    level: DashaLevel
    antardashas: List[AntardashaResponse]


class VimshottariDashaResponse(BaseModel):
    """Complete API response for a Vimshottari Dasha timeline."""
    mahadashas: List[MahadashaResponse]
